luda-adventure/src/app/game/update/calculate_next_motion_of_character/calculate_next_movement_of_character.rs

```diff
@@ -0,0 +1,279 @@
+use crate::{
+    app::game::{
+        known_id::object::PLAYER_CHARACTER, simplify_collision_box_list, zero_speed, Collider,
+        CollisionBox, MovementPath, PlayerCharacter, Positioner, TileExt, Velocity,
+    },
+    ecs,
+};
+use float_cmp::{ApproxEq, F32Margin};
+use namui::prelude::*;
+
+const COLLISION_MARGIN: F32Margin = F32Margin {
+    ulps: 1,
+    epsilon: 0.0001,
+};
+
+pub fn calculate_next_movement_of_character(ecs_app: &mut ecs::App, max_duration: Time) {
+    let target_collider_list = ecs_app.query_entities::<(&Collider, &Positioner)>();
+    let target_collision_box_list = simplify_collision_box_list(
+        target_collider_list
+            .iter()
+            .filter_map(|(target_entity, (target_collider, target_positioner))| {
+                if target_entity.id() == PLAYER_CHARACTER {
+                    None
+                } else {
+                    Some(target_collider.get_collision_box(target_positioner.xy(0.ms())))
+                }
+            })
+            .collect(),
+    );
+
+    if let Some((_character_entity, (character, character_collider, character_positioner))) =
+        ecs_app
+            .query_entities_mut::<(&PlayerCharacter, &Collider, &mut Positioner)>()
+            .first_mut()
+    {
+        let current_time = character_positioner.movement_end_time();
+        let character_xy = character_positioner.xy(current_time);
+        let mut velocity = character.velocity();
+        let mut end_time = current_time + max_duration;
+        let character_collision_box = character_collider.get_collision_box(character_xy);
+
+        resolve_collision(
+            &mut velocity,
+            &mut end_time,
+            current_time,
+            character_collision_box,
+            &target_collision_box_list,
+        );
+
+        let end_time = match predict_remaining_collision_start_time(
+            velocity,
+            character_collision_box,
+            &target_collision_box_list,
+        ) {
+            Some(remaining_collision_start_time) => {
+                Time::min(end_time, current_time + remaining_collision_start_time)
+            }
+            None => end_time,
+        };
+        let delta_time = end_time - current_time;
+        let delta_xy = Xy {
+            x: velocity.x * delta_time,
+            y: velocity.y * delta_time,
+        };
+        let end_xy = character_xy + delta_xy;
+        let next_movement_path = MovementPath::new(current_time, end_time, character_xy, end_xy);
+
+        character_positioner.push_movement_path(next_movement_path);
```

Comment:
Does this system push movement to character positioner without clear? it will make out-of-memory problem and performance decrease for iteration.

Reply:
It is flushed when user input is changed.
Outdated.